driver/others/memory.c

```diff
@@ -415,23 +417,72 @@ int openblas_get_num_threads(void) {
 #endif
 }
 
-struct release_t {
-  void *address;
-  void (*func)(struct release_t *);
-  long attr;
-};
-
 int hugetlb_allocated = 0;
 
 #if defined(OS_WINDOWS)
 #define THREAD_LOCAL __declspec(thread)
-#define UNLIKELY_TO_BE_ZERO(x) (x)
+#define LIKELY_ONE(x) (x)
 #else
 #define THREAD_LOCAL __thread
-#define UNLIKELY_TO_BE_ZERO(x) (__builtin_expect(x, 0))
+#define LIKELY_ONE(x) (__builtin_expect(x, 1))
+#endif
+
+struct alloc_t {
+  int used;
+  int attr;
+  void (*func)(struct alloc_t *);
+  /* Pad to 64-byte alignment */
+#ifdef __64BIT__
+  char pad[48];
+#else
+  char pad[52];
+#endif
+};
+
+static const int allocation_block_size = BUFFER_SIZE + sizeof(struct alloc_t);
+
+/* TLS is supported from clang 2.8, gcc 4.1, MSVC 2005, and XCode 8 */
+#if ((defined(__clang__) && (__clang_major__ > 2) ||               \
+      ((__clang_minor__ == 2) || (__clang_minor__ == 8))) ||       \
+     (!defined(__clang__) && defined(__GNUC__) &&                  \
+      (__GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 1))) || \
+     (defined(_MSC_VER) && _MSC_VER >= 1400)) &&                   \
+    (!defined(__apple_build_version__) || __apple_build_version__ >= 8000042)
+#define HAS_COMPILER_TLS
+#endif
+
+/* Android NDK's before 12 did not support TLS */
+#if defined(__ANDROID__) && defined(__clang__)
+#if __has_include(<android/ndk-version.h>)
+#include <android/ndk-version.h>
+#endif
+#if defined(__ANDROID__) && defined(__clang__) && defined(__NDK_MAJOR__) && \
+    defined(__NDK_MINOR__) &&                                               \
+    ((__NDK_MAJOR__ < 12) || ((__NDK_MAJOR__ == 12) && (__NDK_MINOR__ < 1)))
+#undef HAS_COMPILER_TLS
+#endif
+#endif
+
+/* Holds pointers to allocated memory */
+#if defined(SMP) && !defined(USE_OPENMP)
+#define MAX_ALLOCATING_THREADS MAX_CPU_NUMBER * 2 * MAX_PARALLEL_NUMBER
+static int next_memory_table_pos = 0;
+#ifdef HAS_COMPILER_TLS
+/* Use compiler generated thread-local-storage */
+static int THREAD_LOCAL local_memory_table_pos = 0;
+#else
+/* Use system-dependent thread-local-storage */
+#ifdef OS_WINDOWS
+static DWORD local_storage_key;
+#else
+static pthread_key_t local_storage_key;
+#endif
```

Comment:
since the else and endifs are so nested, consider adding 

#endif // foo to indicate what is ending